File: core/src/de/project/ice/ecs/systems/ControlSystem.java
package de.project.ice.ecs.systems;

import com.badlogic.ashley.core.Entity;
import com.badlogic.ashley.utils.ImmutableArray;
import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.Input;
import com.badlogic.gdx.InputProcessor;
import com.badlogic.gdx.graphics.OrthographicCamera;
import com.badlogic.gdx.math.Rectangle;
import com.badlogic.gdx.math.Vector2;
import com.badlogic.gdx.math.Vector3;
import de.project.ice.ecs.Components;
import de.project.ice.ecs.Families;
import de.project.ice.ecs.IceEngine;
import de.project.ice.ecs.components.*;
import de.project.ice.hotspot.Hotspot;
import de.project.ice.hotspot.Hotspots;
import de.project.ice.inventory.Inventory;
import de.project.ice.pathlib.PathArea;
import de.project.ice.pathlib.PathCalculator;
import de.project.ice.screens.CursorScreen;

import static de.project.ice.config.Config.PIXELS_TO_METRES;

public class ControlSystem extends IteratingIceSystem implements InputProcessor
{
    private ImmutableArray<Entity> cameras;
    private ImmutableArray<Entity> hotspots;
    OrthographicCamera active_camera = null;
    Hotspot active_hotspot = null;
    public Entity hotspot_entity = null;

    private Vector2 pointerPos = new Vector2();
    private boolean pointerDown = false;
    private boolean pointerWasDown = false;
    private boolean mouseDown = false;

    private PathCalculator mouseCalculator = new PathCalculator(0f);

    public CursorScreen.Cursor primaryCursor = CursorScreen.Cursor.None;
    public CursorScreen.Cursor secondaryCursor = CursorScreen.Cursor.None;
    public Inventory.Item active_item = null;
    private IceEngine engine;

    @SuppressWarnings("unchecked")
    public ControlSystem()
    {
        super(Families.controllable);

    }

    @Override
    public void processEntity(Entity entity, float deltaTime)
    {
        TextureComponent texture = Components.texture.get(entity);
        TransformComponent transform = Components.transform.get(entity);

        if (mouseDown)
        {
            entity.remove(UseComponent.class);
            if (Gdx.input.isButtonPressed(Input.Buttons.LEFT) || Gdx.input.isButtonPressed(Input.Buttons.RIGHT))
            {
                if (Gdx.input.isButtonPressed(Input.Buttons.RIGHT) && active_item != null)
                {
                    active_item = null;
                }

                activateHotspot(entity);
                switch (getActiveCursor())
                {
                    case Walk:
                    case Take:
                    case Look:
                    case Speak:

                        float width = PIXELS_TO_METRES * (texture.getRegion().getData() != null ? texture.getRegion().getData().getRegionWidth() : 0);

                        Vector2 target;
                        if (hotspot_entity != null)
                        {
                            HotspotComponent hotspotComponent = Components.hotspot.get(hotspot_entity);
                            target = Components.transform.get(hotspot_entity).getPos().cpy()
                                    .add(hotspotComponent.getOrigin()).add(hotspotComponent.getTargetPos());
                        }
                        else
                        {
                            Vector3 mouse_target = active_camera.unproject(new Vector3(pointerPos.x, pointerPos.y, 0f)); // unprojects UI coordinates to camera coordinates
                            target = new Vector2(mouse_target.x, mouse_target.y);
                        }

                        Vector2 start = transform.getPos().cpy();
                        target = target.sub(width / 2, 0);

                        PathPlanningComponent pathPlanningComponent = engine.createComponent(PathPlanningComponent.class);
                        pathPlanningComponent.setTarget(target);
                        pathPlanningComponent.setStart(start);

                        entity.add(pathPlanningComponent);
                        break;
                }
            }
        }
    }

    private void activateHotspot(Entity entity)
    {
        if (active_hotspot != null)
        {
            UseComponent useComponent = engine.createComponent(UseComponent.class);
            useComponent.setTarget(hotspot_entity);
            useComponent.setCursor(getActiveCursor());
            useComponent.setWithItem(active_item);
            entity.add(useComponent);
        }
        active_item = null;
    }

    private CursorScreen.Cursor getActiveCursor()
    {
        return Gdx.input.isButtonPressed(Input.Buttons.LEFT) || secondaryCursor == CursorScreen.Cursor.None ? primaryCursor : secondaryCursor;
    }

    @Override
    public void update(float deltaTime)
    {
        // update the active camera
        if (cameras.size() > 0)
        {
            active_camera = cameras.first().getComponent(CameraComponent.class).getCamera();
        }

        if (active_camera == null)
        {
            return;
        }

        if (pointerDown && !pointerWasDown)
        {
            mouseDown = true;
        }

        super.update(deltaTime);
        mouseDown = false;
        pointerWasDown = pointerDown;
    }

    @Override
    public void addedToEngine(IceEngine engine)
    {
        super.addedToEngine(engine);
        this.engine = engine;
        cameras = engine.getEntitiesFor(Families.camera);
        hotspots = engine.getEntitiesFor(Families.hotspot);
    }

    @Override
    public boolean keyDown(int keycode)
    {
        return false;
    }

    @Override
    public boolean keyUp(int keycode)
    {
        return false;
    }

    @Override
    public boolean keyTyped(char character)
    {
        return false;
    }

    @Override
    public boolean touchDown(int screenX, int screenY, int pointer, int button)
    {
        pointerPos.set(screenX, screenY);
        pointerDown = true;
        return false;
    }

    @Override
    public boolean touchUp(int screenX, int screenY, int pointer, int button)
    {
        pointerPos.set(screenX, screenY);
        pointerDown = false;
        return false;
    }

    @Override
    public boolean touchDragged(int screenX, int screenY, int pointer)
    {
        return false;
    }

    @Override
    public boolean mouseMoved(int screenX, int screenY)
    {
        pointerPos.set(screenX, screenY);
        primaryCursor = CursorScreen.Cursor.None;
        secondaryCursor = CursorScreen.Cursor.None;
        active_hotspot = null;
        hotspot_entity = null;

        if (active_camera != null)
        {
            Vector3 coords = active_camera.unproject(new Vector3(screenX, screenY, 0f));

            PathArea walkarea = engine.getPathSystem().getWalkArea();
            if (mouseCalculator.IsInside(walkarea, new Vector2(coords.x, coords.y)))
            {
                primaryCursor = CursorScreen.Cursor.Walk;
            }

            for (Entity entity : hotspots)
            {
                TransformComponent transform = Components.transform.get(entity);
                HotspotComponent hotspot = Components.hotspot.get(entity);

                Vector2 pos = transform.getPos().cpy().add(hotspot.getOrigin());
                Vector2 origin = pos.cpy().add(hotspot.getOrigin());

                if (new Rectangle(origin.x, origin.y, hotspot.getWidth(), hotspot.getHeight()).contains(coords.x, coords.y))
                {
                    active_hotspot = Hotspots.INSTANCE.get(hotspot.getScript());
                    if (active_hotspot != null)
                    {
                        hotspot_entity = entity;
                        primaryCursor = active_hotspot.getPrimaryCursor();
                        secondaryCursor = active_hotspot.getSecondaryCursor();
                    }
                    break;
                }
            }
        }
        return true;
    }

    @Override
    public boolean scrolled(int amount)
    {
        return false;
    }
}

File: core/src/de/project/ice/ecs/systems/ControlSystem.kt
package de.project.ice.ecs.systems

import com.badlogic.ashley.core.Entity
import com.badlogic.ashley.utils.ImmutableArray
import com.badlogic.gdx.Gdx
import com.badlogic.gdx.Input
import com.badlogic.gdx.InputProcessor
import com.badlogic.gdx.graphics.OrthographicCamera
import com.badlogic.gdx.math.Rectangle
import com.badlogic.gdx.math.Vector2
import com.badlogic.gdx.math.Vector3
import de.project.ice.ecs.Components
import de.project.ice.ecs.Families
import de.project.ice.ecs.IceEngine
import de.project.ice.ecs.components.*
import de.project.ice.hotspot.Hotspot
import de.project.ice.inventory.Inventory
import de.project.ice.pathlib.PathCalculator
import de.project.ice.screens.CursorScreen
import de.project.ice.ecs.getComponent

import de.project.ice.config.Config.PIXELS_TO_METRES
import de.project.ice.utils.plus

class ControlSystem
@SuppressWarnings("unchecked")
constructor() : IteratingIceSystem(Families.controllable), InputProcessor {
    private var cameras: ImmutableArray<Entity>? = null
    private var hotspots: ImmutableArray<Entity>? = null
    private var controllables: ImmutableArray<Entity>? = null
    internal var active_camera: OrthographicCamera? = null
    internal var active_hotspot: Hotspot? = null
    var hotspot_entity: Entity? = null

    private val pointerPos = Vector2()
    private var pointerDown = false
    private var pointerWasDown = false
    private var mouseDown = false

    private val mouseCalculator = PathCalculator(0f)

    var primaryCursor: CursorScreen.Cursor = CursorScreen.Cursor.None
    var secondaryCursor: CursorScreen.Cursor = CursorScreen.Cursor.None
    var active_item: Inventory.Item? = null
    var cursorText: String = ""
    override fun setProcessing(processing: Boolean) {
        super.setProcessing(processing)
    }

    public override fun processEntity(entity: Entity, deltaTime: Float) {
        val texture = Components.texture.get(entity)
        val transform = Components.transform.get(entity)

        if (mouseDown) {
            entity.remove(UseComponent::class.java)
            if (Gdx.input.isButtonPressed(Input.Buttons.LEFT) || Gdx.input.isButtonPressed(Input.Buttons.RIGHT)) {
                if (Gdx.input.isButtonPressed(Input.Buttons.RIGHT) && active_item != null) {
                    active_item = null
                }

                activateHotspot(entity)
                if (activeCursor!= CursorScreen.Cursor.None) {

                    var target: Vector2
                    if (hotspot_entity != null) {
                        val hotspotComponent = Components.hotspot.get(hotspot_entity)
                        target = Components.transform.get(hotspot_entity).pos + hotspotComponent.origin + hotspotComponent.targetPos
                    } else {
                        val mouse_target = active_camera!!.unproject(Vector3(pointerPos.x, pointerPos.y, 0f)) // unprojects UI coordinates to camera coordinates
                        target = Vector2(mouse_target.x, mouse_target.y)
                    }

                    val start = transform.pos.cpy()

                    val pathPlanningComponent = engine.createComponent(PathPlanningComponent::class.java)
                    pathPlanningComponent.target = target
                    pathPlanningComponent.start = start

                    entity.add(pathPlanningComponent)


                }
            }
        }
    }

    private fun activateHotspot(entity: Entity?) {
        if (active_hotspot != null) {
            val useComponent = engine!!.createComponent(UseComponent::class.java)
            useComponent.target = hotspot_entity
            useComponent.cursor = activeCursor
            useComponent.withItem = active_item
            entity?.add(useComponent)
        }
        active_item = null
    }

    private val activeCursor: CursorScreen.Cursor
        get() = if (Gdx.input.isButtonPressed(Input.Buttons.LEFT) || secondaryCursor === CursorScreen.Cursor.None) primaryCursor else secondaryCursor

    override fun update(deltaTime: Float) {
        // update the active camera
        if (cameras!!.size() > 0) {
            active_camera = Components.camera.get(cameras!!.first()).camera
        }

        if (active_camera == null) {
            return
        }

        if (pointerDown && !pointerWasDown) {
            mouseDown = true
        }

        super.update(deltaTime)

        if (controllables!!.size() == 0 && mouseDown && active_hotspot != null) {
            activateHotspot(hotspot_entity)
        }

        mouseDown = false
        pointerWasDown = pointerDown
    }

    override fun addedToEngine(engine: IceEngine) {
        super.addedToEngine(engine)
        cameras = engine.getEntitiesFor(Families.camera)
        hotspots = engine.getEntitiesFor(Families.hotspot)
        controllables = engine.getEntitiesFor(Families.controllable)
    }

    override fun keyDown(keycode: Int): Boolean {
        return false
    }

    override fun keyUp(keycode: Int): Boolean {
        return false
    }

    override fun keyTyped(character: Char): Boolean {
        return false
    }

    override fun touchDown(screenX: Int, screenY: Int, pointer: Int, button: Int): Boolean {
        pointerPos.set(screenX.toFloat(), screenY.toFloat())
        pointerDown = true
        return false
    }

    override fun touchUp(screenX: Int, screenY: Int, pointer: Int, button: Int): Boolean {
        pointerPos.set(screenX.toFloat(), screenY.toFloat())
        pointerDown = false
        return false
    }

    override fun touchDragged(screenX: Int, screenY: Int, pointer: Int): Boolean {
        return false
    }

    override fun mouseMoved(screenX: Int, screenY: Int): Boolean {
        pointerPos.set(screenX.toFloat(), screenY.toFloat())
        primaryCursor = CursorScreen.Cursor.None
        secondaryCursor = CursorScreen.Cursor.None
        active_hotspot = null
        hotspot_entity = null

        if (active_camera != null) {
            val coords = active_camera!!.unproject(Vector3(screenX.toFloat(), screenY.toFloat(), 0f))

            val walkarea = engine!!.pathSystem.walkArea
            if (mouseCalculator.IsInside(walkarea, Vector2(coords.x, coords.y))) {
                primaryCursor = CursorScreen.Cursor.Walk
            }

            val entity = hotspots?.filter {
                val transform = it.getComponent(Components.transform)
                val hotspot = it.getComponent(Components.hotspot)

                val pos = transform.pos + hotspot.origin
                val origin = pos + hotspot.origin

                if (Rectangle(origin.x, origin.y, hotspot.width, hotspot.height).contains(coords.x, coords.y)) {
                    return@filter true
                }
                return@filter false
            }?.sortedBy {
                it.getComponent(Components.transform)?.z?:0
            }?.firstOrNull()
            if (entity != null) {
                val hotspot = entity.getComponent(Components.hotspot)

                active_hotspot = Hotspot[hotspot.script]
                if (active_hotspot != null) {
                    val text: String
                    try {
                        text = engine.game.strings["${hotspot.cursorText}_text"]
                    } catch (ex: Exception) {
                        text = active_hotspot?.id ?: ""
                    }
                    hotspot_entity = entity
                    primaryCursor = hotspot.primaryCursor
                    secondaryCursor = hotspot.secondaryCursor
                    cursorText = text
                }
            } else {
                cursorText = ""
            }
        }
        return true
    }

    override fun scrolled(amount: Int): Boolean {
        return false
    }
}